1721: Feature/drag and drop favorites

abhithory opened the pull request:
Fixed #1721 
This PR adds Feature to drag and drop your favorites.

What i have done:
- added reorder list in frontend in Favorites component. From that reorder list I am getting dragging element ID and index where it is dropped
- Added backend function for rendering that element with id to a new index
- Added new mutation in Favorite to change the order of favorites
- Added function in useFavorites to change the order
- Added draggable icon in NavItem



Here is the video you can take a look

[Screencast from 08-10-23 02:08:46 PM CDT.webm](https://github.com/twentyhq/twenty/assets/76877003/746bdbc3-6bdd-4731-bd0a-1bb168e312bb)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/twentyhq/twenty?pullRequest=1930) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/twentyhq/twenty?pullRequest=1930) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/twentyhq/twenty?pullRequest=1930) it.</sub>

thaisguigon commented on server/src/core/favorite/resolvers/favorite.resolver.ts:
```diff
@@ -120,6 +130,62 @@ export class FavoriteResolver {
+      where: {
+        workspaceId: workspace.id,
+      },
+      include: {
+        person: true,
+        company: {
+          include: {
+            accountOwner: true,
+          },
+        },
+      },
+    });
+    const currentIndex = allFavorites.findIndex(
+      (favorite) => favorite.id === favoriteId,
+    );
+
+    assert(
+      currentIndex !== -1,
+      'Favorite Item with this Id not found',
+      NotFoundException,
+    );
+    // Remove 'favoriteId' favorite from its current position
+    const FavoriteItemToReorder = allFavorites.splice(currentIndex, 1)[0];
+
+    // Insert 'favoriteId' favorite at the 'toIndex'
+    allFavorites.splice(toIndex, 0, FavoriteItemToReorder);
+    // Delete all old favorites
+    await this.favoriteService.deleteMany({
+      where: { workspaceId: workspace.id },
+    });
+
+    const updatedFavoritesData = allFavorites.map((favorite) => ({
+      workspaceId: workspace.id,
+      ...(favorite?.personId ? { personId: favorite.personId } : {}),
+      ...(favorite?.companyId ? { companyId: favorite.companyId } : {}),
+    }));
+
+    await this.favoriteService.createMany({
+      data: updatedFavoritesData,
+    });
```
Not sure `findMany` will always return the correct order without `orderBy`. Instead of removing + re-creating all favorites, could you:
- add an `index` property to the `Favorite` model in `schema.prisma`
- generate a prisma migration for this with `yarn prisma:migrate`
- create a `updateOneFavorite` resolver (instead of `updateFavoritesOrder`), that will use `this.favoriteService.update`
- then on reorder, all changed favorites index can be updated one by one, and the `useGetFavoritesQuery` hook can use `orderBy: { index: SortOrder.Asc }`

This is how we manage ordering in backend (for instance for the `ViewField` model).

abhithory replied:
ok sure, i will do this.

abhithory replied:
i added index in schema.prisma and then migrated using yarn prisma:migrate but it is now showing this error when i add new Favorite.

`` The column `favorites.index` does not exist in the current database.
Error: 
Invalid `prisma.favorite.findFirst()` invocation: ``

Schema:

`` model Favorite {
  id    String @id @default(uuid())
  index Int?   @default(0)

  workspaceId       String?
  /// @TypeGraphQL.omit(input: true, output: false)
  person            Person?          @relation(fields: [personId], references: [id], onDelete: Cascade)
  personId          String?
  /// @TypeGraphQL.omit(input: true, output: false)
  company           Company?         @relation(fields: [companyId], references: [id], onDelete: Cascade)
  companyId         String?
  /// @TypeGraphQL.omit(input: true, output: false)
  workspaceMember   WorkspaceMember? @relation(fields: [workspaceMemberId], references: [id])
  workspaceMemberId String?

  @@map("favorites")
}
 ``

abhithory replied:
@thaisguigon @charlesBochet , can you please help me in this?

abhithory replied:
@thaisguigon  @charlesBochet  , can you please help me with this?

thaisguigon commented on front/src/modules/favorites/components/Favorites.tsx:
```diff
@@ -1,59 +1,100 @@
+
+  const [favorites, setFavorites] = useState(data?.findFavorites);
+
+  useEffect(() => {
+    const _favorites = data?.findFavorites;
+    const _finalFav = _favorites ? Array.from(_favorites) : [];
+    setFavorites(_finalFav);
+  }, [data?.findFavorites]);
```
Hi @abhithory, thank you for contributing!

You don't need a `useEffect`, instead you can use the `onCompleted` callback like this:

```typescript
  const [favorites, setFavorites] = useState<
    GetFavoritesQuery['findFavorites']
  >([]);

  useGetFavoritesQuery({
    onCompleted: (data) => setFavorites(data?.findFavorites ?? []),
  });
```

abhithory replied:
ok,

abhithory commented on front/src/modules/favorites/components/Favorites.tsx:
```diff
@@ -1,59 +1,100 @@
-          )) ??
-          (company && (
-            <NavItem
+      <Reorder.Group
+        axis="y"
+        values={favorites}
+        onReorder={setFavorites}
+        as="div"
+      >
```
yes definitely, i will them. Thanks

abhithory commented:
Hi @charlesBochet, I have updated the backend and frontend according to @thaisguigon description.

Can you please help me with these problems?
Everything else is working fine but here are two problems:
- I have changed the schema of Favorite Modal in `schema.prisma` and then run yarn prisma:migrate but after that when I am making any new Company or Person Favorite.

Error: 
The column `favorites.index` does not exist in the current database. Error:  Invalid `prisma.favorite.findFirst()` invocation:

=> I manually added column index in the database and then tested

- And second problem is that The `WORKSPACE` section comes up and then goes down when i drag and drop items. You can see this in video
-

[Screencast from 15-10-23 12:59:35 PM CDT.webm](https://github.com/twentyhq/twenty/assets/76877003/b2e7ab88-b97f-4f86-89c5-eba10bfbba3f)

Bonapara commented:
Thank you, @abhithory, for the pull request. From a design standpoint:

- You can eliminate the grip icon adjacent to the favorite icon. 
- You may introduce a focus state with a ` transparent:light` background color. It should span the full width of the menu item container, similar to the hover state.
Add a 2px high blue line to mark the item's new position after reordering. This line should resemble the one used for column resizing.

Take a look at what Notion.so is doing for an illustration:

https://github.com/twentyhq/twenty/assets/19412894/5a76d493-6682-49f8-b243-303736c14d9f

Thank you 🙏

abhithory commented:
Ok @Bonapara, I will do this. Can you please help me with the backend issue that i mentioned above?

Bonapara commented:
> Ok @Bonapara, I will do this. Can you please help me with the backend issue that i mentioned above?

I can't, I'm the product guy here ;)

charlesBochet commented:
And this one too @abhithory I will do it tomorrow, I promise!

abhithory commented:
Hello @charlesBochet , @Bonapara . I have created a new PR for this: #2097 . Please check that.

This was with old code, thats why there were some small bugs, that are fixed now.

charlesBochet commented:
Ok, I'm taking a look right now

abhithory commented:
ok @charlesBochet , Please look at new PR #2097
